In [1]:
import pandas as pd
import numpy as np

In [2]:
# import datasets
admissions = pd.read_csv(r"C:\Users\akann\OneDrive\Desktop\AMDARI\Project 3 - Predictive Healthcare Capacity Management\MEDOPTIX-ANALYTICS\Dataset-medoptix\admissions.csv")
daily_metrics = pd.read_csv(r"C:\Users\akann\OneDrive\Desktop\AMDARI\Project 3 - Predictive Healthcare Capacity Management\MEDOPTIX-ANALYTICS\Dataset-medoptix\daily_metrics.csv")
hospitals = pd.read_csv(r"C:\Users\akann\OneDrive\Desktop\AMDARI\Project 3 - Predictive Healthcare Capacity Management\MEDOPTIX-ANALYTICS\Dataset-medoptix\hospitals.csv")
main_admissions = pd.read_csv(r"C:\Users\akann\OneDrive\Desktop\AMDARI\Project 3 - Predictive Healthcare Capacity Management\MEDOPTIX-ANALYTICS\Dataset-medoptix\main_admissions.csv")
main_daily_metrics = pd.read_csv(r"C:\Users\akann\OneDrive\Desktop\AMDARI\Project 3 - Predictive Healthcare Capacity Management\MEDOPTIX-ANALYTICS\Dataset-medoptix\main_daily_metrics.csv")
wards = pd.read_csv(r"C:\Users\akann\OneDrive\Desktop\AMDARI\Project 3 - Predictive Healthcare Capacity Management\MEDOPTIX-ANALYTICS\Dataset-medoptix\wards.csv")

#### Preview of each Dataset

In [5]:
datasets = {
    "admissions": admissions,
    "daily_metrics": daily_metrics,
    "hospitals": hospitals,
    "main_admissions": main_admissions,
    "main_daily_metrics": main_daily_metrics,
    "wards": wards
}

for name, df in datasets.items():
    print(f"\n--- {name.upper()} (first 5 rows) ---")
    print(df.head(5))



--- ADMISSIONS (first 5 rows) ---
   admission_id        date  hospital_id ward_code arrival_source  \
0                1  2022-10-28            1        ED       referral   
1                2  2022-10-28            1       ICU      ambulance   
2                3  2022-10-28            1        ED           self   
3                4  2022-10-28            1       MED           self   
4                5  2022-10-28            1        ED           self   

  triage_level  wait_minutes  length_of_stay_days      outcome  age sex  \
0          4.0           189                    0   discharged   48   M   
1          4.0           189                    0  transferred   93   F   
2          NaN           189                    0   discharged   52   M   
3          3.0           189                    0   discharged   64   M   
4          4.0           189                    0   discharged   59   M   

   procedure_flag  
0             0.0  
1             0.0  
2             0.0  
3

In [ ]:
# Column Names for each Dataset

datasets = {
    "admissions": admissions,
    "daily_metrics": daily_metrics,
    "hospitals": hospitals,
    "main_admissions": main_admissions,
    "main_daily_metrics": main_daily_metrics,
    "wards": wards
}

for name, df in datasets.items():
    print(f"\n--- {name.upper()} COLUMN NAMES ---")
    print(df.columns.tolist())



--- ADMISSIONS COLUMN NAMES ---
['admission_id', 'date', 'hospital_id', 'ward_code', 'arrival_source', 'triage_level', 'wait_minutes', 'length_of_stay_days', 'outcome', 'age', 'sex', 'procedure_flag']

--- DAILY_METRICS COLUMN NAMES ---
['date', 'hospital_id', 'ward_code', 'base_beds', 'effective_capacity', 'occupancy', 'overflow', 'admissions', 'discharges', 'staffing_index', 'avg_wait_minutes', 'admission_rate_per_bed', 'discharge_rate_per_bed', 'occupancy_rate']

--- HOSPITALS COLUMN NAMES ---
['hospital_id', 'hospital_name', 'city', 'country', 'latitude', 'longitude', 'founded_year']

--- MAIN_ADMISSIONS COLUMN NAMES ---
['admission_id', 'date', 'hospital_id', 'ward_code', 'arrival_source', 'triage_level', 'wait_minutes', 'length_of_stay_days', 'outcome', 'age', 'sex', 'procedure_flag']

--- MAIN_DAILY_METRICS COLUMN NAMES ---
['date', 'hospital_id', 'ward_code', 'base_beds', 'effective_capacity', 'occupancy', 'overflow', 'admissions', 'discharges', 'staffing_index', 'avg_wa

#### Merge Data

In [8]:
admissions = pd.concat([admissions, main_admissions], ignore_index=True)
metrics = pd.concat ([daily_metrics, main_daily_metrics], ignore_index=True)

In [ ]:
# Check duplicate rows

print("Admissions duplicates:", admissions.duplicated().sum())
print("Metrics duplicates:", metrics.duplicated().sum())

Admissions duplicates: 0
Metrics duplicates: 0


In [10]:
admissions

,admission_id,date,hospital_id,ward_code,arrival_source,triage_level,wait_minutes,length_of_stay_days,outcome,age,sex,procedure_flag,admission_id
0,1.0,2022-10-28,1,ED,referral,4.0,189,0,discharged,48,M,0.0,NaN
1,2.0,2022-10-28,1,ICU,ambulance,4.0,189,0,transferred,93,F,0.0,NaN
2,3.0,2022-10-28,1,ED,self,NaN,189,0,discharged,52,M,0.0,NaN
3,4.0,2022-10-28,1,MED,self,3.0,189,0,discharged,64,M,0.0,NaN
4,5.0,2022-10-28,1,ED,self,4.0,189,0,discharged,59,M,0.0,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
169271,NaN,2024-03-17,5,SURG,self,NaN,104,1,discharged,65,F,1.0,189393.0
169272,NaN,2024-03-17,5,SURG,self,NaN,104,3,discharged,60,F,1.0,189394.0
169273,NaN,2024-03-17,5,SURG,self,NaN,104,5,readmit_30d,81,M,1.0,189395.0
169274,NaN,2024-03-18,5,SURG,ambulance,NaN,105,3,discharged,65,M,1.0,189396.0


In [11]:
metrics

,date,hospital_id,ward_code,base_beds,effective_capacity,occupancy,overflow,admissions,discharges,staffing_index,avg_wait_minutes,admission_rate_per_bed,discharge_rate_per_bed,occupancy_rate,date
0,2022-10-28,1,MED,30,34,34,26,51,17,0.999,219,1.700,0.567,1.000,NaN
1,2022-10-29,1,SURCG,30,34,34,35,62,31,0.938,223,2.067,1.033,1.000,NaN
2,2022-10-30,1,ICU,30,34,34,39,63,28,0.953,275,2.100,0.933,1.000,NaN
3,2022-10-31,1,ED,30,34,34,26,56,34,0.974,227,1.867,1.133,1.000,NaN
4,2022-11-01,1,MED,30,34,34,29,55,30,1.013,257,1.833,1.000,1.000,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21968,NaN,5,SURG,40,46,46,15,26,17,0.944,197,0.650,0.425,1.000,2025-10-22
21969,NaN,5,SURG,40,46,46,9,20,17,0.962,214,0.500,0.425,1.000,2025-10-23
21970,NaN,5,SURG,40,46,43,3,15,18,0.957,154,0.375,0.450,0.935,2025-10-24
21971,NaN,5,SURG,40,40,34,0,9,18,0.908,32,0.225,0.450,0.850,2025-10-25


#### Check for the inconsistencies

In [12]:
list_data = [
    ("admissions", admissions), 
    ("metrics", metrics), 
    ("hospitals", hospitals), 
    ("wards", wards)
]

for name, data in list_data:
    print(name)
    print(f"{round(data.isnull().mean() * 100, 2)}\n")

    print("Duplicates:", data.duplicated().sum())

    print("Numeric Features Summary:")
    print(data.describe())

    print("Categorical / Object Features Summary:")
    print(data.describe(include=("object", "category")))

    print("-" * 100)

admissions
admission_id        99.68
date                    0.00
hospital_id             0.00
ward_code               0.00
arrival_source          0.00
triage_level           68.63
wait_minutes            0.00
length_of_stay_days     0.00
outcome                 0.00
age                     0.00
sex                     0.00
procedure_flag          0.00
admission_id            0.32
dtype: float64

Duplicates: 0
Numeric Features Summary:
       admission_id    hospital_id   wait_minutes  length_of_stay_days  \
count        545.00000  169276.000000  169276.000000        169276.000000   
mean         273.00000       2.861197     192.670266             3.026460   
std          157.47222       1.448333      75.111417             3.070374   
min            1.00000       1.000000     -15.000000            -5.000000   
25%          137.00000       2.000000     142.000000             0.000000   
50%          273.00000       3.000000     184.000000             2.000000   
75%          409.

#### Calculate Outliers using Empirical Rule and z-score
** The what, why and the when

##### Empirical Rule (What, why and when)
- What it is:
Assumes data is normally distributed.
Anything outside mean ± 3 standard deviations is considered an outlier.

- When to use it:
When data is roughly bell-shaped (normal distribution).
When you want a quick, simple outlier check.
When the dataset is large and the distribution is stable

- Why use it:
Fast and easy.
Good for variables like heights, lab values, many biological measures.
Works well if data follows a normal pattern.

##### Z-Score Method (What, why and when)
- What it is:
Measures how many standard deviations a value is from the mean.

- When to use it:
When there is need for more precision than the empirical rule.
When the distribution is approximately normal but not perfect.
When comparing multiple variables on different scales.
To score how extreme each point is, not just flag it.

- Why use it:
Works even if the distribution is not perfectly normal.
Gives exact "extremeness" of each value (useful in modeling).
More flexible (threshold can be adjusted)



#### Calculate Outlier using Empirical Rule

In [ ]:
# Drop NaN values and include percentage (%) of the outliers

for name, data in list_data:
    print(f"\n=== OUTLIER SUMMARY: {name} ===")

    for col in data.select_dtypes(include='number'):
        col_data = data[col].dropna()
        
        mean = col_data.mean()
        std = col_data.std()
        
        lower = mean - 3*std
        upper = mean + 3*std
        
        outliers = data[(data[col] < lower) | (data[col] > upper)]
        
        percent = (len(outliers) / len(data)) * 100
        
        print(f"{col}: {len(outliers)} outliers ({percent:.2f}%)")
    
    print('_' * 70)



=== OUTLIER SUMMARY: admissions ===
admission_id: 0 outliers (0.00%)
hospital_id: 0 outliers (0.00%)
wait_minutes: 1393 outliers (0.82%)
length_of_stay_days: 918 outliers (0.54%)
age: 230 outliers (0.14%)
procedure_flag: 1 outliers (0.00%)
admission_id: 0 outliers (0.00%)
______________________________________________________________________

=== OUTLIER SUMMARY: metrics ===
hospital_id: 0 outliers (0.00%)
base_beds: 0 outliers (0.00%)
effective_capacity: 0 outliers (0.00%)
occupancy: 0 outliers (0.00%)
overflow: 123 outliers (0.56%)
admissions: 40 outliers (0.18%)
discharges: 0 outliers (0.00%)
staffing_index: 23 outliers (0.10%)
avg_wait_minutes: 132 outliers (0.60%)
admission_rate_per_bed: 96 outliers (0.44%)
discharge_rate_per_bed: 0 outliers (0.00%)
occupancy_rate: 792 outliers (3.60%)
______________________________________________________________________

=== OUTLIER SUMMARY: hospitals ===
hospital_id: 0 outliers (0.00%)
latitude: 0 outliers (0.00%)
longitude: 0 outliers (0.0

#### Calculate Ouliers using Z-score for evry numeric column in each dataset

In [17]:
for name, data in list_data:
    print(f"\n=== Z-SCORE OUTLIER SUMMARY: {name} ===")

    numeric_cols = data.select_dtypes(include='number').columns

    for col in numeric_cols:
        col_data = data[col].dropna()
        mean = col_data.mean()
        std = col_data.std()

        if std == 0:
            print(f"{col}: STD = 0 → Skipped")
            print('-' * 50)
            continue

        z = (col_data - mean) / std
        outliers = col_data[abs(z) > 3]

        print(f"Column: {col}")
        print(f"Outlier Count: {len(outliers)}")
        print('-' * 50)

    print('_' * 70)



=== Z-SCORE OUTLIER SUMMARY: admissions ===
Column: admission_id
Outlier Count: 0
--------------------------------------------------
Column: hospital_id
Outlier Count: 0
--------------------------------------------------
Column: wait_minutes
Outlier Count: 1393
--------------------------------------------------
Column: length_of_stay_days
Outlier Count: 918
--------------------------------------------------
Column: age
Outlier Count: 230
--------------------------------------------------
Column: procedure_flag
Outlier Count: 1
--------------------------------------------------
Column: admission_id
Outlier Count: 0
--------------------------------------------------
______________________________________________________________________

=== Z-SCORE OUTLIER SUMMARY: metrics ===
Column: hospital_id
Outlier Count: 0
--------------------------------------------------
Column: base_beds
Outlier Count: 0
--------------------------------------------------
Column: effective_capacity
Outlier C

# Insights from Outlier Analysis

1. Admissions Dataset: outliers are mainly in: wait_minutes (1393), length_of_stay_days (918) & age (230)
- This shows high variability in hospitals (long waits, long stays, unusual ages).

2. Metrics Dataset: outliers are in: occupancy_rate (792), avg_wait_minutes (132) & admission_rate_per_bed (96)
- This indicates days with extreme crowding, staffing imbalance, or high demand.
3. No outliers in Hospitals & Wards.

#### Drop irrelevant columns

In [21]:
cols_to_drop = ['admission_id', 'date']

for name, data in list_data:
    print(name)
    for col in cols_to_drop:
        if col in data.columns:
            data.drop(col, axis=1, inplace=True)

admissions
metrics
hospitals
wards


In [22]:
admissions.columns

Index(['date', 'hospital_id', 'ward_code', 'arrival_source', 'triage_level',
       'wait_minutes', 'length_of_stay_days', 'outcome', 'age', 'sex',
       'procedure_flag', 'admission_id'],
      dtype='object')

#### Fill null value
** The what (to understand the meaning i.e. what it does), The why (why is it important) and the when (under which situation it is important)

- forward fill (works well with times series data)
- backward fill ( same as this)
- grouped mode imputation 
- KNN imputer
- Interpolation

#### Deal with the outliers

** The what (to understand the meaning i.e. what it does), The why (why is it important) and the when (under which situation it is important)

- Winsorization
- Scaling/log transformation

In [23]:
admissions['triage_level'].value_counts()

triage_level
4.0    84757
3.0    21069
4.0    15812
2.0     7897
5.0     5217
1.0     2566
3.0      214
2.0       76
5.0       69
1.0       28
6.5        1
abc        1
Name: count, dtype: int64

In [ ]:
values_to_replace ={
    "6.5": None,
    "abc" : None
}

admissions['triage_level'].replace(values_to_replace, inplace= True)

admissions['triage_level'].value_counts()

C:\Users\akann\AppData\Local\Temp\ipykernel_16760\319161237.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  admissions['triage_level'].replace(values_to_replace, inplace= True)


triage_level
4.0    84757
3.0    21069
4.0    15812
2.0     7897
5.0     5217
1.0     2566
3.0      214
2.0       76
5.0       69
1.0       28
Name: count, dtype: int64

In [26]:
admissions['arrival_source'].nunique()

4

In [27]:
cols_to_fill = ['triage_level', 'data']

for name, data in list_data:
    for col in cols_to_fill:
        if 'triage_level' in data.columns:
            data['triage_level'] = data.groupby('ward_code')['triage_level'].transform(lambda x: x.fillna(x.mode()))

            if 'date' in data.columns:
                data['data'] = pd.to_datetime(data['date'], errors = 'coerce')
                data['date'] = data['date'].fillna(method='ffill')
                data['date'] = data['date'].fillna(method='bfill')


C:\Users\akann\AppData\Local\Temp\ipykernel_16760\1080141750.py:10: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['date'] = data['date'].fillna(method='ffill')
C:\Users\akann\AppData\Local\Temp\ipykernel_16760\1080141750.py:11: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['date'] = data['date'].fillna(method='bfill')
C:\Users\akann\AppData\Local\Temp\ipykernel_16760\1080141750.py:10: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['date'] = data['date'].fillna(method='ffill')
C:\Users\akann\AppData\Local\Temp\ipykernel_16760\1080141750.py:11: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['date'] = data['date'].fillna(

#### Handling Outliers using Winsorization Technique

In [28]:
for name, data in list_data:
    for col in data.select_dtypes(include='number').columns.to_list():

        lower = data[col].quantile(0.05)
        upper = data[col].quantile(0.95)

        data[col].clip(lower=lower, upper=upper, inplace=True)

C:\Users\akann\AppData\Local\Temp\ipykernel_16760\3892972364.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[col].clip(lower=lower, upper=upper, inplace=True)
C:\Users\akann\AppData\Local\Temp\ipykernel_16760\3892972364.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

F

In [29]:
# Run this again to confirm null values has been handled

for name, data in list_data:
    print(name)
    for col in data.select_dtypes(include='number').columns.to_list():
        mean = data[col].mean()
        std = data[col].std()

        lower = mean - (3 * std)
        upper = mean + (3 * std)

        outliers = data[(data[col] < lower) | (data[col] > upper)]

        print(col)
        print(len(outliers))
        print('- * 50')
    print('_' * 50)

admissions
hospital_id
0
- * 50
wait_minutes
0
- * 50
length_of_stay_days
0
- * 50
age
0
- * 50
procedure_flag
0
- * 50
admission_id
0
- * 50
__________________________________________________
metrics
hospital_id
0
- * 50
base_beds
0
- * 50
effective_capacity
0
- * 50
occupancy
0
- * 50
overflow
0
- * 50
admissions
0
- * 50
discharges
0
- * 50
staffing_index
0
- * 50
avg_wait_minutes
0
- * 50
admission_rate_per_bed
0
- * 50
discharge_rate_per_bed
0
- * 50
occupancy_rate
1161
- * 50
__________________________________________________
hospitals
hospital_id
0
- * 50
latitude
0
- * 50
longitude
0
- * 50
founded_year
0
- * 50
__________________________________________________
wards
hospital_id
0
- * 50
base_beds
0
- * 50
__________________________________________________


In [31]:
admissions['arrival_source'].unique()

array(['referral', 'ambulance', 'self', 'transfer'], dtype=object)

In [34]:
for name, data in list_data:
    for col in data.select_dtypes(include=['object', 'category']).columns.to_list():
        print(data[col].value_counts())
        print('_' * 50)

date
2022-11-27      180
2022-10-30      180
2022-11-02      180
2022-11-01      180
2022-11-18      180
               ... 
2025-07-08      139
2024-07-07      139
2025-09-13      137
2024-07-28      133
invalid-date      1
Name: count, Length: 1096, dtype: int64
__________________________________________________
ward_code
ED       53062
MED      43818
ICU      40814
SURG     31569
SURCG       13
Name: count, dtype: int64
__________________________________________________
arrival_source
self         101509
ambulance     30671
referral      25299
transfer      11797
Name: count, dtype: int64
__________________________________________________
triage_level
3.0    21068
4.0    15812
2.0     7897
5.0     5217
1.0     2566
3.0      214
4.0      155
2.0       76
5.0       69
1.0       28
Name: count, dtype: int64
__________________________________________________
outcome
discharged     159360
readmit_30d      4451
transferred      4150
death            1314
unknown             1
Name: count,

#### Merging data

In [32]:
admissions

,date,hospital_id,ward_code,arrival_source,triage_level,wait_minutes,length_of_stay_days,outcome,age,sex,procedure_flag,admission_id,data
0,2022-10-28,1,ED,referral,4.0,189,0,discharged,48,M,0.0,NaN,2022-10-28
1,2022-10-28,1,ICU,ambulance,4.0,189,0,transferred,86,F,0.0,NaN,2022-10-28
2,2022-10-28,1,ED,self,3.0,189,0,discharged,52,M,0.0,NaN,2022-10-28
3,2022-10-28,1,MED,self,3.0,189,0,discharged,64,M,0.0,NaN,2022-10-28
4,2022-10-28,1,ED,self,4.0,189,0,discharged,59,M,0.0,NaN,2022-10-28
...,...,...,...,...,...,...,...,...,...,...,...,...,...
169271,2024-03-17,5,SURG,self,NaN,104,1,discharged,65,F,1.0,180960.5,2024-03-17
169272,2024-03-17,5,SURG,self,NaN,104,3,discharged,60,F,1.0,180960.5,2024-03-17
169273,2024-03-17,5,SURG,self,NaN,104,5,readmit_30d,81,M,1.0,180960.5,2024-03-17
169274,2024-03-18,5,SURG,ambulance,NaN,105,3,discharged,65,M,1.0,180960.5,2024-03-18


In [33]:
metrics

,hospital_id,ward_code,base_beds,effective_capacity,occupancy,overflow,admissions,discharges,staffing_index,avg_wait_minutes,admission_rate_per_bed,discharge_rate_per_bed,occupancy_rate,date
0,1,MED,30,34,34,26,51,17,0.999,219,1.700,0.567,1.000,NaN
1,1,SURCG,30,34,34,35,62,31,0.938,223,2.067,1.033,1.000,NaN
2,1,ICU,30,34,34,39,63,28,0.953,275,2.100,0.933,1.000,NaN
3,1,ED,30,34,34,26,56,33,0.974,227,1.867,1.100,1.000,NaN
4,1,MED,30,34,34,29,55,30,1.013,257,1.833,1.000,1.000,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21968,5,SURG,40,46,46,15,26,17,0.944,197,0.650,0.425,1.000,2025-10-22
21969,5,SURG,40,46,46,9,20,17,0.962,214,0.500,0.425,1.000,2025-10-23
21970,5,SURG,40,46,43,3,15,18,0.957,154,0.375,0.450,0.935,2025-10-24
21971,5,SURG,40,40,34,1,9,18,0.909,89,0.325,0.450,0.935,2025-10-25


## Task 2
Merge immediately: state your reason for that and how you are going about analyzing as a single data.
Not merge immediately: state your reason. What do we do? How do we merge? What feature do we engineer from the admissions to match the metrcs table?

In [34]:
# Aggregate numerical columns

admissions.dtypes 

date                           object
hospital_id                     int64
ward_code                      object
arrival_source                 object
triage_level                   object
wait_minutes                    int64
length_of_stay_days             int64
outcome                        object
age                             int64
sex                            object
procedure_flag                float64
admission_id                  float64
data                   datetime64[ns]
dtype: object

In [35]:
admissions.select_dtypes(include='number') 

,hospital_id,wait_minutes,length_of_stay_days,age,procedure_flag,admission_id
0,1,189,0,48,0.0,NaN
1,1,189,0,86,0.0,NaN
2,1,189,0,52,0.0,NaN
3,1,189,0,64,0.0,NaN
4,1,189,0,59,0.0,NaN
...,...,...,...,...,...,...
169271,5,104,1,65,1.0,180960.5
169272,5,104,3,60,1.0,180960.5
169273,5,104,5,81,1.0,180960.5
169274,5,105,3,65,1.0,180960.5


In [36]:
admission_agg = admissions.groupby(
    ['date', 'hospital_id', 'ward_code']
).agg({
    'wait_minutes': 'mean',
    'length_of_stay_days': 'mean',
    'age': 'mean',
    'procedure_flag': 'sum'
}).reset_index()


In [37]:
admission_agg

,date,hospital_id,ward_code,wait_minutes,length_of_stay_days,age,procedure_flag
0,2022-10-28,1,ED,183.333333,0.000,53.111111,0.0
1,2022-10-28,1,ICU,107.400000,5.000,56.100000,0.0
2,2022-10-28,1,MED,117.000000,3.100,61.400000,2.0
3,2022-10-28,1,SURCG,189.000000,0.000,45.500000,0.0
4,2022-10-28,1,SURG,87.000000,2.375,61.000000,5.0
...,...,...,...,...,...,...,...
19284,2025-10-26,4,SURG,87.000000,3.000,47.750000,5.0
19285,2025-10-26,5,ED,207.000000,0.000,58.666667,0.0
19286,2025-10-26,5,ICU,140.000000,7.500,61.333333,0.0
19287,2025-10-26,5,MED,213.000000,3.375,55.500000,1.0


In [ ]:
use technic close to one hot Encoding Warning. 

### Merged Data

In [ ]:

pd.merge(admission_agg, metrics, on=['date', 'hospital_id', 'ward_code'], how='inner')

,date,hospital_id,ward_code,wait_minutes,length_of_stay_days,age,procedure_flag,base_beds,effective_capacity,occupancy,overflow,admissions,discharges,staffing_index,avg_wait_minutes,admission_rate_per_bed,discharge_rate_per_bed,occupancy_rate
0,2022-10-28,1,ED,183.333333,0.000,53.111111,0.0,30,34,34,26,51,17,0.999,219,1.700,0.567,1.000
1,2022-10-28,1,ICU,107.400000,5.000,56.100000,0.0,20,23,23,1,13,7,1.033,89,0.650,0.350,0.935
2,2022-10-28,1,MED,117.000000,3.100,61.400000,2.0,50,58,58,19,50,11,1.009,147,1.000,0.220,1.000
3,2022-10-28,1,SURG,87.000000,2.375,61.000000,5.0,40,40,33,1,15,11,0.951,89,0.375,0.275,0.935
4,2022-10-28,2,ED,331.000000,0.000,55.333333,1.0,30,34,34,42,67,19,0.964,299,2.233,0.633,1.000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
18799,2025-10-26,4,MED,195.000000,4.125,55.500000,0.0,50,58,58,19,34,23,0.951,173,0.680,0.460,1.000
18800,2025-10-26,4,SURG,87.000000,3.000,47.750000,5.0,40,40,38,1,10,16,0.961,89,0.325,0.400,0.950
18801,2025-10-26,5,ED,207.000000,0.000,58.666667,0.0,30,34,34,30,53,27,0.938,228,1.767,0.900,1.000
18802,2025-10-26,5,ICU,140.000000,7.500,61.333333,0.0,20,23,23,2,7,7,0.909,117,0.325,0.350,0.957


In [ ]:
complete the feature engineering

## Task 1
Check the incorrect spellings and convert to null values and handle appropriately

In [38]:
for name, data in list_data:
    print(f"Cleaning {name}...")

    # Ward code typos
    if 'ward_code' in data.columns:
        data['ward_code'] = data['ward_code'].replace({'SURCG': np.nan})

    # Sex column rare/invalid entries
    if 'sex' in data.columns:
        data['sex'] = data['sex'].replace({'X': np.nan})

    # Outcome rare/invalid
    if 'outcome' in data.columns:
        data['outcome'] = data['outcome'].replace({'unknown': np.nan})

    # Dates: convert invalid to NaT
    if 'date' in data.columns:
        data['date'] = pd.to_datetime(data['date'], errors='coerce')

    # triage_level: convert invalid strings to NaN
    if 'triage_level' in data.columns:
        data['triage_level'] = pd.to_numeric(data['triage_level'], errors='coerce')

    print(f"{name} cleaned.\n")


Cleaning admissions...
admissions cleaned.

Cleaning metrics...
metrics cleaned.

Cleaning hospitals...
hospitals cleaned.

Cleaning wards...
wards cleaned.



#### Handle missing values after cleaning

In [54]:
for name, data in list_data:
    print(f"Handling missing values in {name}...")

    if 'triage_level' in data.columns and 'ward_code' in data.columns:
        data['triage_level'] = data.groupby('ward_code')['triage_level'].transform(
            lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else np.nan)
        )

    # Categorical columns filled with mode
    for col in ['ward_code', 'sex', 'outcome', 'arrival_source']:
        if col in data.columns and data[col].isna().sum() > 0:
            data[col] = data[col].fillna(data[col].mode()[0])

    # Fill missing dates using forward-fill, then backward-fill
    if 'date' in data.columns:
        data['date'] = data['date'].fillna(method='ffill').fillna(method='bfill')

    print(f"Missing values handled for {name}.\n")


Handling missing values in admissions...
Missing values handled for admissions.

Handling missing values in metrics...
Missing values handled for metrics.

Handling missing values in hospitals...
Missing values handled for hospitals.

Handling missing values in wards...
Missing values handled for wards.



C:\Users\akann\AppData\Local\Temp\ipykernel_16760\1314096680.py:16: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['date'] = data['date'].fillna(method='ffill').fillna(method='bfill')


In [55]:
# Check for remaining missing values

for name, data in list_data:
    print(f"{name} missing values summary:\n")
    print(data.isna().sum())
    print("-" * 50)


admissions missing values summary:

date                     0
hospital_id              0
ward_code                0
arrival_source           0
triage_level             0
wait_minutes             0
length_of_stay_days      0
outcome                  0
age                      0
sex                      0
procedure_flag           0
admission_id           545
data                     0
dtype: int64
--------------------------------------------------
metrics missing values summary:

hospital_id               0
ward_code                 0
base_beds                 0
effective_capacity        0
occupancy                 0
overflow                  0
admissions                0
discharges                0
staffing_index            0
avg_wait_minutes          0
admission_rate_per_bed    0
discharge_rate_per_bed    0
occupancy_rate            0
date                      0
dtype: int64
--------------------------------------------------
hospitals missing values summary:

hospital_id      0
hospit

In [56]:
for name, data in list_data:
    print(f"Handling remaining missing values in {name}...")

    # Fill triage_level by ward mode, then overall mode
    if 'triage_level' in data.columns and 'ward_code' in data.columns:
        data['triage_level'] = data.groupby('ward_code')['triage_level'].transform(
            lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else np.nan)
        )
        data['triage_level'] = data['triage_level'].fillna(data['triage_level'].mode()[0])

    # Fill categorical columns with mode
    for col in ['ward_code', 'sex', 'outcome', 'arrival_source']:
        if col in data.columns:
            data[col] = data[col].fillna(data[col].mode()[0])

    # Fill missing dates
    if 'date' in data.columns:
        data['date'] = data['date'].fillna(method='ffill').fillna(method='bfill')

   
    print(f"Missing values handled for {name}.\n")


Handling remaining missing values in admissions...
Missing values handled for admissions.

Handling remaining missing values in metrics...
Missing values handled for metrics.

Handling remaining missing values in hospitals...
Missing values handled for hospitals.

Handling remaining missing values in wards...
Missing values handled for wards.



C:\Users\akann\AppData\Local\Temp\ipykernel_16760\2171240095.py:18: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['date'] = data['date'].fillna(method='ffill').fillna(method='bfill')


In [59]:
# Fill procedure_flag missing value
admissions['procedure_flag'] = admissions['procedure_flag'].fillna(admissions['procedure_flag'].mode()[0])

# Fill 'data' column missing value
admissions['data'] = admissions['data'].fillna(method='ffill').fillna(method='bfill')

# Optional: handle missing admission_id (drop or leave)
admissions = admissions.dropna(subset=['admission_id'])  # only if missing IDs are critical


C:\Users\akann\AppData\Local\Temp\ipykernel_16760\664176661.py:5: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  admissions['data'] = admissions['data'].fillna(method='ffill').fillna(method='bfill')


## Task 2
Merge immediately: state your reason for that and how you are going about analyzing as a single data.
Do not merge immediately: state your reason. What do we do? How do we merge? What feature do we engineer from the admissions to match the metrcs table?

##### Reasons why merging immediately is not an option
We do not merge immediately because the datasets are at different granularities:
1) admissions is patient-level (one row per admission)
2) metrics is daily/ward-level (one row per ward per day).

If merge immediately:
- We might end up duplicating metric rows for multiple admissions per day.
- Any analysis could be misleading, e.g., inflating bed occupancy or admission rates.

##### What to be done before merging
1) Aggregate the admissions data so it is at the same level as metrics: daily + hospital + ward.
2) Engineer meaningful features from admissions to summarize patient-level data.

##### How to merge
Once admissions are aggregated to daily + hospital + ward, merge with metrics using key columns like: date, hospital_id, ward_code.
- This ensures a one-to-one or many-to-one match between admissions and metrics.

##### To engineer features from admissions to match the metrics table. 
We will create aggregated features per day/ward/hospital. These aggregated features night align with the metrics data and allow analysis at the same granularity, such as occupancy, staffing, or patient flow.

In [32]:
for name, data in list_data:
    for col in data.select_dtypes(include=['object', 'category'])
    print(data[col].value_counts())
    print('_' *50)

SyntaxError: expected ':' (988661469.py, line 2)